### PR TITLE
Full re-work of AWS wodle as per wazuh/wazuh#510

### PR DESCRIPTION
**Pull request by UranusBytes**
Full re-work of AWS wodle as per wazuh/wazuh#510
- Scalability and functional enhancements for parsing of CloudTrail
- Support for existing config params
- Upgrade to a granular object key addressing to support multiple CloudTrails in S3 bucket
- Support granular parsing by account id, region, prefix
- Support only parsing logs after a given date
- Support IAM credential profiles, IAM roles
- Only look for new logs/objects since last iteration
- Skip digest files altogether (only look at logs)
- Move from downloading object and working with file on filesystem to byte stream
- Inherit debug from modulesd
- Add bounds checks for msg against socket buffer size; truncate fields if too big (wazuh/wazuh#733)
- Support multiple debug levels
- Move connect error so not confused with general error
- If fail to parse log, and skip_on_error, attempt to send me msg to wazuh
- Support existing configurations by migrating data, inferring other required params

Associated documentation updates covered in wazuh/wazuh-documentation#274

### CODE REVIEW

**Comment by UranusBytes** [comment]
To put some more context around this pull request (started on wazuh/wazuh#510) : 

The existing AWS wodle had the following limitations that were preventing me from using it:
* The wodle would list every single object in the S3 bucket, query the DB to see if it had been parsed, and then if new, try to parse it.  This doesn't scale for s3 buckets that have a large number of objects
  * I have S3 buckets with multiple years of archived CloudTrail logs, as well as other logs (s3 access logs, etc...) 
* The wodle would only support a single s3 bucket
  * I have multiple S3 buckets spread across multiple AWS accounts, with each S3 bucket hosting logs from multiple AWS accounts
* The wodle would only support a single set of IAM credentials (so incompatible with multiple buckets, AWS credential file with multiple profiles, IAM role assumption, etc...)

To try and address these, what I did is:
* Change the wodle so that each cloudtrail could be distinctly defined.  Each defined cloudtrail could have unique settings (bucket, prefix, credentials, etc...) independent of others.  Each defined cloudtrail would spin off its own instance of aws.py script
* Change wodle to support additional parameters such as CloudTrail options (prefix) as well as limiting parsing of logs to only logs after a specific date (prevent parsing of older/legacy logs on first runs) or specific regions
* The aws.py script is more limited in how it lists objects:  it only lists objects within the key namespace of the defined AWS account (so doesn't matter how many other accounts or other objects are in the S3 bucket), if the only_logs_after is defined, it will only look for logs after that date, and if logs for that given AWS account/region have already been parsed, it picks up at the last log.  It does all of this by taking advantage of the MARKER argument of the list_objects_v2 function, and the fact that the key namespace will always be incremental based upon the format/date. 
* Instead of downloading the log object to a file on the file system, the object is kept in memory, decompressed, and parsed from there
* Additional error handling, including multiple levels of debug, inherit of debug from the modulesd, etc...
* Support truncating of excessive log event sizes - wazuh/wazuh#733
* Cleanup/truncate the DB to keep from growing too big/unwieldy

These optimizations, in my opinion, have quite a few advantages:
* The wodle processes at execution time much faster and scales significantly further.  In context, the setup I'm currently testing this on is gathering CloudTrail logs from 3 different S3 buckets supporting logs from 20 AWS accounts that are all very actively used in multiple regions, and the entire wodle normally executes with 30 seconds on a small EC2 instance.  If needed, there's no reason it couldn't be scaled further with parallel execution
* The cost of running the wodle is probably reduced slightly by not trying to list all objects in the s3 bucket at each execution.  Can be further reduced by limiting regions (wouldn't recommend)
* Supports more complex, flexible deployments
* Sets the stage to support additional log collection by the AWS wodle (I'm eyeing S3 access logs and ALB/ELB now)

To support backwards compatibility, it should support existing ossec.conf configurations without issue (in full disclosure, I tried to test all possible permutations on this, but since I couldn't successfully run the old one in my environment, I can't say that I thought of all scenarios).  It would be great to fully deprecate these older settings...

**Comment by vikman90** [comment]
Hi @UranusBytes,

This is a brilliant work! This will add so much value to the integration with AWS.

Please let us do some testing on it and we will write back to you with our results. First of all, I found a little issue that prevents the manager from being compiled. I will mark the line involved in the review.

Thank you very much for your contribution.

Kind regards.

File: src/config/wmodules-config.c [hunk 1]
```diff
@@ -92,14 +84,14 @@ int Read_WModule(const OS_XML *xml, xml_node *node, void *d1, void *d2)
...
     else if (!strcmp(node->values[0], WM_AWS_CONTEXT.name)) {
-        if (wm_aws_read(children, cur_wmodule, agent_cfg) < 0) {
+        if (wm_aws_read(xml, children, cur_wmodule) < 0) {
             OS_ClearNode(children);
             return OS_INVALID;
         }
     }
 #ifndef CLIENT
     else if (!strcmp(node->values[0], WM_VULNDETECTOR_CONTEXT.name)) {
-        if (wm_vulnerability_detector_read(xml, children, cur_wmodule) < 0) {
```
Comment (vikman90) [inline src/config/wmodules-config.c]: This change produced this error when compiling the manager:

```
config/wmodules-config.c: In function ‘Read_WModule’:
config/wmodules-config.c:94:44: warning: passing argument 1 of ‘wm_vulnerability_detector_read’ from incompatible pointer type [-Wincompatible-pointer-types]
         if (wm_vulnerability_detector_read(children, cur_wmodule) < 0) {
                                            ^~~~~~~~
In file included from ./wazuh_modules/wmodules.h:64:0,
                 from config/wmodules-config.c:12:
./wazuh_modules/wm_vuln_detector.h:279:5: note: expected ‘const OS_XML * {aka const struct _OS_XML *}’ but argument is of type ‘xml_node ** {aka struct _xml_node **}’
 int wm_vulnerability_detector_read(const OS_XML *xml, xml_node **nodes, wmodule *module);
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
config/wmodules-config.c:94:54: warning: passing argument 2 of ‘wm_vulnerability_detector_read’ from incompatible pointer type [-Wincompatible-pointer-types]
         if (wm_vulnerability_detector_read(children, cur_wmodule) < 0) {
                                                      ^~~~~~~~~~~
In file included from ./wazuh_modules/wmodules.h:64:0,
                 from config/wmodules-config.c:12:
./wazuh_modules/wm_vuln_detector.h:279:5: note: expected ‘xml_node ** {aka struct _xml_node **}’ but argument is of type ‘wmodule * {aka struct wmodule *}’
 int wm_vulnerability_detector_read(const OS_XML *xml, xml_node **nodes, wmodule *module);
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
config/wmodules-config.c:94:13: error: too few arguments to function ‘wm_vulnerability_detector_read’
         if (wm_vulnerability_detector_read(children, cur_wmodule) < 0) {
             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from ./wazuh_modules/wmodules.h:64:0,
                 from config/wmodules-config.c:12:
./wazuh_modules/wm_vuln_detector.h:279:5: note: declared here
 int wm_vulnerability_detector_read(const OS_XML *xml, xml_node **nodes, wmodule *module);
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

The previous version of this line, including the parameter `xml`, lets the manager compile.

**Comment by UranusBytes** [comment]
Thanks @vikman90 

My bad...  I had been working off the v3.3.0 source bundle and forgot to do a diff against master before the pull to see that I wasn't stepping on some other changes from the master branch.  The changes should be localized to src/config/wmodules-config.c and rolled back now...

Please do as much testing as you can!  I've been running this for a couple weeks now without issue, but I can't by any means handle all possible issues/scenarios.  If I had to choose an area that I'm most uncomfortable with, it would be around the migration of legacy user data in the existing log_progress table (since I couldn't get the previous version to run, I could only imitate the scenario for testing).

And as I'm looking thru the code again, I see a bone head mistake with the migration process...  *sigh*

Let me know what I can do to assist with further testing...

**Comment by UranusBytes** [comment]
So I've run into a problem and I'm not sure how to proceed...

Within a CloudTrail log entry, there are 3 fields ('additionalEventData', 'requestParameters', and 'responseElements') that will vary in data structures depending upon the event itself (sometimes they are a string, sometimes they are a nested structure of strings/lists/objects).  This was causing alerts to not be successfully inserted into Elasticsearch by LogStash with errors like this:

> [2018-07-31T23:53:55,137][WARN ][logstash.outputs.elasticsearch] Could not index event to Elasticsearch. {:status=>400, :action=>["index", {:_id=>nil, :_index=>"wazuh-alerts-3.x-2018.07.30", :_type=>"wazuh", :_routing=>nil}, #<LogStash::Event:0x746c56c>], :response=>{"index"=>{"_index"=>"wazuh-alerts-3.x-2018.07.30", "_type"=>"wazuh", "_id"=>"jjzD8mQBnItSooZeLqAb", "status"=>400, "error"=>{"type"=>"mapper_parsing_exception", "reason"=>"object mapping for [data.aws.requestParameters.instanceType] tried to parse field [instanceType] as object, but found a concrete value"}}}}
> [2018-07-31T23:53:57,698][WARN ][logstash.outputs.elasticsearch] Could not index event to Elasticsearch. {:status=>400, :action=>["index", {:_id=>nil, :_index=>"wazuh-alerts-3.x-2018.07.30", :_type=>"wazuh", :_routing=>nil}, #<LogStash::Event:0x36fab2c6>], :response=>{"index"=>{"_index"=>"wazuh-alerts-3.x-2018.07.30", "_type"=>"wazuh", "_id"=>"jzzD8mQBnItSooZeOKoQ", "status"=>400, "error"=>{"type"=>"mapper_parsing_exception", "reason"=>"object mapping for [data.aws.requestParameters.ebsOptimized] tried to parse field [ebsOptimized] as object, but found a concrete value"}}}}

This is a problem that existed with the original codebase; it just probably never got noticed...

Since none of these fields are used by a rule at this time, my thought was to just convert everything under this to a pretty print string of the json.  When the time comes to write some rules against these fields, can just use a nested json decoder...

So in my test environment, I added the following code around line 499 of aws.py
```
# Some fields in CloudTrail are dynamic in nature, which causes problems for ES mapping
for field_to_str in ['additionalEventData', 'requestParameters','responseElements']:
    if field_to_str in aws_log:
        try:
            aws_log[field_to_str] = json.dumps(aws_log[field_to_str],
                                               ensure_ascii=True,
                                               indent=2,
                                               sort_keys=True)
        except:
            aws_log[field_to_str] = 'Value unable to convert to string'
```
The resulting alerts.json looked nice...  But I'm still learning Elasticsearch, so I didn't think of the consequences...

Now the problem I realized is that the existing index conflicts with this new document structure as fed from alerts.json, which results in the majority of alerts not getting into Elasticsearch with errors like:

> [2018-08-03T13:52:52,156][WARN ][logstash.outputs.elasticsearch] Could not index event to Elasticsearch. {:status=>400, :action=>["index", {:_id=>nil, :_index=>"wazuh-alerts-3.x-2018.08.03", :_type=>"wazuh", :_routing=>nil}, #<LogStash::Event:0x67937b2a>], :response=>{"index"=>{"_index"=>"wazuh-alerts-3.x-2018.08.03", "_type"=>"wazuh", "_id"=>"HD4PAGUBnItSooZe-2FK", "status"=>400, "error"=>{"type"=>"mapper_parsing_exception", "reason"=>"object mapping for [data.aws.requestParameters] tried to parse field [requestParameters] as object, but found a concrete value"}}}}

In order to get past this, I can think of 3 solutions:
1. Drop/re-create the ES index, and just not import any previous alerts (not preferred and very disruptive, but easiest path)
2. Do a ES reindex with an ingest pipeline to drop/ignore those three fields from older alerts (not preferred, but at least keeps most of all the old alerts)
3. Drop/re-create the ES index, and then do a re-import of all the alerts.json, re-structuring the json or prune the fields on the fly during the import.  Something akin to the restore_alerts.sh used for 2.x to 3.x upgrade (would keep full alert history, but will be a significant migration process)  

I know I can get my data shifted over sufficiently for me, but I'm trying to think the best way to approach this for other users that would migrate/upgrade to this code.

My assumption is to do ##1 for an upgrade to this code, and then offer a script that would accomplish ##3 by importing all the old alerts and pruning the offending fields for old alerts (something like the 2.x to 3.x restore_alerts.sh) 

Any suggestions on how to proceed?

**Comment by UranusBytes** [comment]
After fighting much of yesterday to try to get the data migrated in ES, I thought of a resolution last night.  By moving data structure namespace (from data.aws.* to data.aws.cloudtrail.event.*), there will be no conflict with Elasticsearch.  The older data will just get deprecated as the older ES indices get pruned.  This also helps me for the other integration I've started poking at: AWS Trusted Advisor...

**Comment by mgmacias95** [comment]
Hello @UranusBytes,

I'm glad you solved that problem. I like the solution you've done. I think it is necessary to update our ruleset in order to support *different kinds of events*: each JSON structure represents a type of event, and those events cannot be decoded the same way 😄 .

Thank you so much for your work.

Best regards,
Marta

File: src/config/wmodules-aws.c [hunk 2]
```diff
@@ -85,48 +100,213 @@ int wm_aws_read(xml_node **nodes, wmodule *module, int agent_cfg)
...
+        } else if (!strcmp(nodes[i]->element, XML_SKIP_ON_ERROR)) {
+            if (!strcmp(nodes[i]->content, "yes"))
+                aws_config->skip_on_error = 1;
+            else if (!strcmp(nodes[i]->content, "no"))
+                aws_config->skip_on_error = 0;
+            else {
+                merror("Invalid content for tag '%s' at module '%s'.", XML_SKIP_ON_ERROR, WM_AWS_CONTEXT.name);
+                return OS_INVALID;
+            }
+        } else if (!strcmp(nodes[i]->element, XML_SKIP_ON_ERROR)) {
```
[the remark this answers is not in the document]
Comment (UranusBytes) [reply src/config/wmodules-aws.c]: Oops...  I had some previous merge problems, and thought I'd caught them all, but obviously I hadn't...

File: wodles/aws/aws.py [hunk 3]
```diff
@@ -5,163 +5,601 @@
...
-                os.remove(downloaded_file_path)
-            except IOError as e:
-                print "ERROR: Cannot delete %s (%s)" % (downloaded_file_path, e.strerror)
+                bucket_files = s3_client.list_objects_v2(**filter_args)
+                if 'Contents' not in bucket_files:
+                    debug('+++ No logs to process: {aws_account_id}/{aws_region}'.format(aws_account_id=aws_account_id,
+                                                                                         aws_region=aws_region), 1)
+                    continue
+
+                for bucket_file in s3_client.list_objects_v2(**filter_args)['Contents']:
```
[the remark this answers is not in the document]
Comment (UranusBytes) [reply wodles/aws/aws.py]: As I was walking thru the code this morning, I was (reluctantly) coming to the same conclusion...

Let me take another stab at it...

File: src/config/wmodules-aws.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (mgmacias95) [reply src/config/wmodules-aws.c]: Don't worry, those things are very common. That's why peer review is necessary 😄 ! I'll let you know if I find something else.

File: wodles/aws/aws.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mgmacias95) [reply wodles/aws/aws.py]: I'm glad we came to the same conclusion 😄 . Let me know if you need something.

**Comment by mgmacias95** [comment]
Hello @UranusBytes,

I've been testing your PR and found out a little bug. I removed the `aws_account_id` option from the configuration and it was validated by `ossec-logtest` but when running the woodle no error was shown. Several improvements need to be added:

* If the `aws_account_id` must be present in the configuration. Return `OS_INVALID` if it's not present. Here you can see an example of that:
https://github.com/wazuh/wazuh/blob/458d330c3e2dfb186f85754292f6a5a7ac7ddbd2/src/config/cluster-config.c#L40-L42
* This is what I've seen when I was doing the test:
   * Setting debug level to INFO:
   ```
   2018/08/10 05:07:05 wazuh-modulesd:aws-cloudtrail: INFO: Starting fetching of logs.
   2018/08/10 05:07:05 wazuh-modulesd:aws-cloudtrail: WARNING: CloudTrail:  -  Returned exit code 2
   2018/08/10 05:07:05 wazuh-modulesd:aws-cloudtrail: WARNING: CloudTrail:  -  SIGINT received
   2018/08/10 05:07:05 wazuh-modulesd:aws-cloudtrail: INFO: Fetching logs finished.
   ```
   * Setting debug level to DEBUG2:
   ```
   2018/08/10 05:16:15 wazuh-modulesd:aws-cloudtrail: DEBUG: Found a cloudtrail tag
   2018/08/10 05:16:15 wazuh-modulesd:aws-cloudtrail: DEBUG: Creating first cloudtrail structure
   2018/08/10 05:16:15 wazuh-modulesd:aws-cloudtrail: DEBUG: Loop thru child nodes
   2018/08/10 05:16:15 wazuh-modulesd:aws-cloudtrail: DEBUG: Parse child node: bucket
   2018/08/10 05:16:15 wazuh-modulesd:aws-cloudtrail: DEBUG: Parse child node: access_key
   2018/08/10 05:16:15 wazuh-modulesd:aws-cloudtrail: DEBUG: Parse child node: secret_key
   2018/08/10 05:16:16 wazuh-modulesd:aws-cloudtrail: INFO: Module AWS started
   2018/08/10 05:16:16 wazuh-modulesd:aws-cloudtrail: INFO: Starting fetching of logs.
   2018/08/10 05:16:16 wazuh-modulesd:aws-cloudtrail: DEBUG: Executing CloudTrail: LEGACY - wazuh-events
   2018/08/10 05:16:16 wazuh-modulesd:aws-cloudtrail: DEBUG: Create argument list
   2018/08/10 05:16:16 wazuh-modulesd:aws-cloudtrail: DEBUG: Launching CloudTrail Command: /var/ossec/wodles/aws/aws.py --bucket wazuh-events --access_key xxxxxx --secret_key xxxx --debug 2 --skip_on_error
   2018/08/10 05:16:16 wazuh-modulesd:aws-cloudtrail: WARNING: CloudTrail:  -  Returned exit code 2
   2018/08/10 05:16:16 wazuh-modulesd:aws-cloudtrail: WARNING: CloudTrail:  -  SIGINT received
   2018/08/10 05:16:16 wazuh-modulesd:aws-cloudtrail: INFO: Fetching logs finished.
   ```

   To exactly know what was happening, I had to run the cloudtrail command:
   ```
   # /var/ossec/wodles/aws/aws.py --bucket wazuh-events --access_key xxxx--secret_key xxxx --debug 2 --skip_on_error
   usage: usage: aws.py [options]
   aws.py: error: argument -c/--aws_account_id is required
   ```
   Wouldn't it be better to show that error in the ossec.log?

I'll continue the testing and I'll send you more feedback 😄 . Thank you so much for doing this!!!

Best regards,
Marta

**Comment by mgmacias95** [comment]
Hi @UranusBytes,

I've been testing the PR and it works perfectly! Thank you so much for all this work, it will add lots of value to Wazuh. 

We want to include this in our next release, this is why I'll merge this in the dev branch [`dev-3.5-aws-cloudtrail-510`](https://github.com/wazuh/wazuh/tree/dev-aws-cloudtrail-510) and continue the development (for example, what I told you yesterday) there.

Don't hesitate to reach me if you need something 😄 .

Best regards,
Marta

**Pull request merged**